internal/util/function/rerank/model_function.go

```diff
@@ -0,0 +1,384 @@
+/*
+ * # Licensed to the LF AI & Data foundation under one
+ * # or more contributor license agreements. See the NOTICE file
+ * # distributed with this work for additional information
+ * # regarding copyright ownership. The ASF licenses this file
+ * # to you under the Apache License, Version 2.0 (the
+ * # "License"); you may not use this file except in compliance
+ * # with the License. You may obtain a copy of the License at
+ * #
+ * #     http://www.apache.org/licenses/LICENSE-2.0
+ * #
+ * # Unless required by applicable law or agreed to in writing, software
+ * # distributed under the License is distributed on an "AS IS" BASIS,
+ * # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * # See the License for the specific language governing permissions and
+ * # limitations under the License.
+ */
+
+package rerank
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/milvus-io/milvus-proto/go-api/v2/commonpb"
+	"github.com/milvus-io/milvus-proto/go-api/v2/schemapb"
+	"github.com/milvus-io/milvus/internal/util/function"
+	"github.com/milvus-io/milvus/internal/util/function/models/utils"
+	"github.com/milvus-io/milvus/pkg/v2/util/paramtable"
+)
+
+const (
+	providerParamName string = "provider"
+	vllmProviderName  string = "vllm"
+	teiProviderName   string = "tei"
+
+	queryKeyName    string = "queries"
+	maxBatchKeyName string = "max_batch"
+)
+
+type modelProvider interface {
+	rerank(context.Context, string, []string) ([]float32, error)
+	getURL() string
+}
+
+type baseModel struct {
+	url      string
+	maxBatch int
+
+	queryKey string
+	docKey   string
+
+	parseScores func([]byte) ([]float32, error)
+}
+
+func (base *baseModel) getURL() string {
+	return base.url
+}
+
+func (base *baseModel) rerank(ctx context.Context, query string, docs []string) ([]float32, error) {
+	requestBodies, err := genRerankRequestBody(query, docs, base.maxBatch, base.queryKey, base.docKey)
+	if err != nil {
+		return nil, err
+	}
+	scores := []float32{}
+	for _, requestBody := range requestBodies {
+		rerankResp, err := base.callService(ctx, requestBody, 30)
+		if err != nil {
+			return nil, fmt.Errorf("Call rerank model failed: %v\n", err)
+		}
+		scores = append(scores, rerankResp...)
+	}
+
+	if len(scores) != len(docs) {
+		return nil, fmt.Errorf("Call Rerank service failed, %d docs but got %d scores", len(docs), len(scores))
+	}
+	return scores, nil
+}
+
+func (base *baseModel) callService(ctx context.Context, requestBody []byte, timeoutSec int64) ([]float32, error) {
+	ctx, cancel := context.WithTimeout(ctx, time.Duration(timeoutSec)*time.Second)
+	defer cancel()
+	headers := map[string]string{
+		"Content-Type": "application/json",
+	}
+	body, err := utils.RetrySend(ctx, requestBody, http.MethodPost, base.url, headers, 3, 1)
```

Comment:
Let's add an exponential sleep time for retry

Reply:
fixed